lib/evmone/instructions_calls.cpp

```diff
@@ -95,6 +96,21 @@ Result call_impl(StackTop stack, int64_t gas_left, ExecutionState& state) noexce
     if (has_value && intx::be::load<uint256>(state.host.get_balance(state.msg->recipient)) < value)
         return {EVMC_SUCCESS, gas_left};  // "Light" failure.
 
+    if constexpr (Op == OP_DELEGATECALL)
+    {
+        if (state.rev >= EVMC_CANCUN && is_eof_container(state.original_code))
+        {
+            // The code targeted by DELEGATECALL must also be an EOF.
+            // This restriction has been added to EIP-3540 in
+            // https://github.com/ethereum/EIPs/pull/6359
+            uint8_t target_code_prefix[2];
+            const auto s = state.host.copy_code(
+                msg.code_address, 0, target_code_prefix, std::size(target_code_prefix));
+            if (!is_eof_container({target_code_prefix, s}))
+                return {EVMC_SUCCESS, gas_left};
```

Comment:
Hm, why is this a success?

Reply:
Success here means we don't revert the current call frame. But the instruction still returns `0`.

Reply:
Ah you mean because we already push `0` on the stack in the beginning, and we only overwrite that on L100 in case of "actual success".

Reply:
Yep. Common pattern used here for some other checks too.